network/local.go

```diff
@@ -0,0 +1,459 @@
+
+	baseUID uint64
+}
+
+// NewLocalContext returns a fresh new context that can be used by LocalConn,
+// LocalListener & LocalHost.
+func NewLocalContext() *LocalContext {
+	return &LocalContext{
+		queues:    make(map[endpoint]*connQueue),
+		listening: make(map[Address]func(Conn)),
+	}
+}
+
+var defaultLocalContext = NewLocalContext()
+
+// endpoint represents one endpoint of a connection.
+type endpoint struct {
+	addr Address
+	// uid is a unique identifier of the remote endpoint
+	// it's unique  for each direction:
+	// 127.0.0.1:2000 -> 127.0.0.1:2000 => 14
+	// 127.0.0.1:2000 <- 127.0.0.1:2000 => 15
+	uid uint64
+}
+
+// LocalReset reset the whole map of connections + listener so it is like
+// a fresh defaultLocalContext.
+func LocalReset() {
+	defaultLocalContext = NewLocalContext()
+
+}
+
+// IsListening returns true if the remote address is listening "virtually"
+func (ccc *LocalContext) IsListening(remote Address) bool {
+	ccc.Lock()
+	defer ccc.Unlock()
+	_, ok := ccc.listening[remote]
+	return ok
+}
+
+// Listening put the address as "listening" mode. If a user connects to this
+// addr, this function will be called.
+func (ccc *LocalContext) Listening(addr Address, fn func(Conn)) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	ccc.listening[addr] = fn
+}
+
+// StopListening remove the address from the "listening" mode
+func (ccc *LocalContext) StopListening(addr Address) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	delete(ccc.listening, addr)
+}
+
+// Connect will check if the remote address is listening, if yes it creates
+// the two connections, and launch the listening function in a go routine.
+// It returns the outgoing connection with any error.
+func (ccc *LocalContext) Connect(local, remote Address) (*LocalConn, error) {
+	ccc.Lock()
+	defer ccc.Unlock()
+
+	fn, ok := ccc.listening[remote]
+	if !ok {
+		return nil, fmt.Errorf("%s can't connect to %s: it's not listening", local, remote)
+	}
+
+	outEndpoint := endpoint{local, ccc.baseUID}
+	ccc.baseUID++
+
+	incEndpoint := endpoint{remote, ccc.baseUID}
+	ccc.baseUID++
+
+	outgoing := newLocalConn(ccc, outEndpoint, incEndpoint)
+	incoming := newLocalConn(ccc, incEndpoint, outEndpoint)
+
+	// outgoing knows how to store packet into the incoming's queue
+	ccc.queues[outEndpoint] = outgoing.connQueue
+	// incoming knows how to store packet into the outgoing's queue
+	ccc.queues[incEndpoint] = incoming.connQueue
+
+	go fn(incoming)
+	return outgoing, nil
+}
+
+// Send will get the connection denoted by this endpoint and will call queueMsg
+// with the packet as argument on it. It returns ErrClosed if it does not find
+// the connection.
+func (ccc *LocalContext) Send(e endpoint, nm Packet) error {
+	ccc.Lock()
+	defer ccc.Unlock()
+	q, ok := ccc.queues[e]
+	if !ok {
+		return ErrClosed
+	}
+
+	q.Push(nm)
+	return nil
+}
+
+// Close will get the connection denoted by this endpoint and will Close it if
+// present.
+func (ccc *LocalContext) Close(conn *LocalConn) {
```

Comment:
If it's a feature that we want to have, I prefer having it at the Connection level than here. 
Here, with two connections `c1` and `c2`, `c1.Close` calls `localContext.Close(c1)` calls `c2.Close` which calls `localContext.Close(c2)` which terminate s (the `if !ok { return }` part). So a close on a connection directly closes its remote counterpart. 
So `if the connection is already closed` means in our `LocalContext` that we are closing the remote counterpart already, so it's not an error. 
But we can definitely impl that features in the `LocalConn` itself, would that be OK ?